src/mock/ms8.rs

```diff
@@ -0,0 +1,815 @@
+    // Source: https://en.wikipedia.org/wiki/Modular_exponentiation#Right-to-left_binary_method
+    if modulus == 1 {
+        return 0;
+    }
+
+    // Need to use 64 bits to ensure the assert from Schneier's algorithm:
+    // (modulus - 1) * (modulus - 1) does not overflow base
+    let mut result: u64 = 1;
+    let md: u64 = u64::from(modulus);
+    let mut base = u64::from(base) % md;
+
+    while exp > 0 {
+        if exp % 2 == 1 {
+            result = (result * base) % md;
+        }
+        exp >>= 1;
+        base = (base * base) % md;
+    }
+
+    result as u32
+}
+
+/// Eigth mersenne prime, aka `i32::MAX`.
+pub const MS8: u32 = 2_147_483_647;
+
+/// Eighth mersenne prime field element
+///
+/// Finite field of order `2^31-1`.
+#[derive(Copy, Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Mersenne8(pub u32);
+
+// We opt to implement the same variants the standard library implements as well:
+//
+// FooAssign: `lhs .= rhs, lhs .= &rhs`.
+// Foo: `lhs . rhs, &lhs . rhs, &lhs . &rhs, lhs . &rhs`
+macro_rules! generate_op_impls {
+    ($t:ty, $op_name:ident, $op_name_assign:ident, $op_method:ident, $op_method_assign:ident) => {
+        // Supplied by caller: `lhs .= rhs`
+
+        // `lhs . rhs`
+        impl ops::$op_name for $t {
+            type Output = Self;
+
+            #[inline]
+            fn $op_method(mut self, rhs: $t) -> Self {
+                self.$op_method_assign(&rhs);
+                self
+            }
+        }
+
+        // lhs .= &rhs
+        impl<'a> ops::$op_name_assign<&'a $t> for $t {
+            #[inline]
+            fn $op_method_assign(&mut self, rhs: &'a $t) {
+                ops::$op_name_assign::$op_method_assign(self, *rhs)
+            }
+        }
+
+        // `&lhs . rhs`
+        impl<'a> ops::$op_name<$t> for &'a $t {
+            type Output = $t;
+
+            #[inline]
+            fn $op_method(self, other: $t) -> Self::Output {
+                let mut tmp = *self;
+                tmp.$op_method_assign(&other);
+                tmp
+            }
+        }
+
+        // `&lhs . &rhs`
+        impl<'a, 'b> ops::$op_name<&'b $t> for &'a $t {
+            type Output = $t;
+
+            #[inline]
+            fn $op_method(self, other: &'b $t) -> Self::Output {
+                let mut tmp = *self;
+                tmp.$op_method_assign(other);
+                tmp
+            }
+        }
+
+        // `lhs . &rhs`
+        impl<'a> ops::$op_name<&'a $t> for $t {
+            type Output = $t;
+
+            #[inline]
+            fn $op_method(mut self, other: &'a $t) -> Self::Output {
+                self.$op_method_assign(other);
+                self
+            }
+        }
+    };
+}
+
+impl ops::SubAssign<Mersenne8> for Mersenne8 {
+    #[inline]
+    fn sub_assign(&mut self, rhs: Mersenne8) {
+        // Since `self.0` is < 2^31-1, `self.0 + 2^31-1` is still smaller than `u32::MAX`.
+        self.0 = (self.0 + MS8 - rhs.0) % MS8;
```

Comment:
I would assume that's within the realm of the compiler to optimize. `MS8` is constant, it would be similar to using a bit-shift to multiply by 2, would it not?

Reply:
I'd be surprised if it recognized the special case of modulo _2<sup>n</sup> - 1_. But then, the compiler keeps surprising me.
Anyway, certainly something we can try out later.